core/server/proxy/src/main/java/alluxio/proxy/s3/S3ObjectTask.java

```diff
@@ -0,0 +1,1346 @@
+        PropertyKey.PROXY_S3_COMPLETE_MULTIPART_UPLOAD_KEEPALIVE_ENABLED);
+        private final Long mKeepAliveTime = Configuration.getMs(
+        PropertyKey.PROXY_S3_COMPLETE_MULTIPART_UPLOAD_KEEPALIVE_TIME_INTERVAL);
+
+        public CompleteMultipartUploadTask(S3Handler handler, OpType opType) {
+            super(handler, opType);
+        }
+
+        @Override
+        public void handleTaskAsync() {
+            try {
+                final String user = mHandler.getUser();
+                mBucket = mHandler.getBucket();
+                mObject = mHandler.getObject();
+                final String uploadId = mHandler.getQueryParameter("uploadId");
+                LOG.debug("(bucket: {}, object: {}, uploadId: {}) queuing task...",
+                        mBucket, mObject, uploadId);
+                HttpServletResponse httpServletResponse = mHandler.getServletResponse();
+
+                // Set headers before getting committed when flushing whitespaces
+                httpServletResponse.setContentType(MediaType.APPLICATION_XML);
+
+                CompletableFuture<Response> respFut = new CompletableFuture<>();
+                S3RequestServlet.esHeavy_.submit(() -> {
+                    Response completeMpUploadResponse = mHandler.getS3Task().continueTask();
+                    respFut.complete(completeMpUploadResponse);
+                });
+                if (mKeepAliveEnabled) {
+                    // Set status before getting committed when flushing whitespaces
+                    httpServletResponse.setStatus(HttpServletResponse.SC_OK);
+                    long sleepMs = 1000;
+                    while (!respFut.isDone()) {
+                        LOG.debug("(bucket: {}, object: {}, uploadId: {}) sleeping for {}ms...",
+                                mBucket, mObject, uploadId, sleepMs);
+                        try {
+                            Thread.sleep(sleepMs);
+                        } catch (InterruptedException e) {
+                            LOG.error(e.toString());
+                        }
+                        // TODO(czhu): figure out how to send whitespace characters while still
+                        // returning a correct status code
+                        // - calling getWriter().flush() commits the response (headers, status code, etc.)
+                        // - https://docs.oracle.com/javaee/7/api/javax/servlet/ServletResponse.html#getWriter--
+                        // periodically sends white space characters to keep the connection from timing out
+                        LOG.debug("(bucket: {}, object: {}, uploadId: {}) sending whitespace...",
+                                mBucket, mObject, uploadId);
+                        httpServletResponse.getWriter().print(" ");
+                        httpServletResponse.getWriter().flush();
+                        sleepMs = Math.min(2 * sleepMs, mKeepAliveTime);
```

Comment:
Digression: it's a little strange here, the way to wait future looks no different from a sync task. And the time even may cost more than a sync one... maybe we can consider refactor it in the future.

Reply:
what do u mean? this is the same code copied from completemultipartuploadhandler, same with that, it is waiting on the actual task happening in another thread to complete, while maintaining the conversation by sending down white spaces, i didnt change this logic, just fit the logic into this architecture so we don't have a separate completempuploadhdlr

Reply:
yeah, I know it's the old code. just for reminding. I think we could have a better solution for it. maybe after this PR.